src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,411 @@
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs = { 0 };
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+
+	memset(&cs, 0, sizeof(cs));
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	cs.version = atoi(p);
+	p = strtokm(NULL, "$");
+	cs.algo_id = atoi(p);
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_IV_PREFIX_SIZE; i++)
+		cs.iv[i] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Init AES CTR counter
+	cs.iv[ENC_IV_SIZE - 1] = 1;
+
+	p = strtokm(NULL, "$");
+	for (i = 0; i < ENC_SIG_SIZE; i++)
+		cs.encrypted_data[i + 4] =
+		    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i +
+		            1])];
+
+	// Keychain for version 3
+	if (cs.version == 3) {
+		p = strtokm(NULL, "$");
+		for (i = 0; i < ENC_KEYCHAIN_SIZE; i++)
+			cs.keychain[i] =
+			    (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 *
+			            i + 1])];
+	}
+
+	MEM_FREE(keeptr);
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+	int nb_keys;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
+	if (1 == cur_salt->version) {
+		nb_keys = 1 << (cur_salt->algo_id - 1);
+	} else {
+		nb_keys = 8;
+	}
+
+#ifdef _OPENMP
+	#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
+		int j;
+		MD5_CTX ctx;
+		unsigned char kdf_out[8][ENC_KEY_SIZE];
+		unsigned char ivs[8][AES_BLOCK_SIZE] = { 0 };
+		unsigned char hash[ENC_KEY_SIZE];
+		unsigned char result[ENC_KEY_SIZE * 8] = { 0 };
+		unsigned char *result_ptr;
+
+		// Key derivation
+		MD5_Init(&ctx);
+		MD5_Update(&ctx, saved_key[index], strlen(saved_key[index]));
+		MD5_Final(hash, &ctx);
+
+		for (j = 0; j < nb_keys; j++) {
+			memcpy(kdf_out[j], default_salts[j], ENC_SALT_SIZE);
+		}
+
+		for (j = 1; j < ENC_DEFAULT_NUM_HASH_ITERATIONS; j++) {
+			MD5_Init(&ctx);
+			MD5_Update(&ctx, hash, 16);
+			MD5_Final(hash, &ctx);
+			enc_xor_block((uint64_t *) result, (uint64_t *) hash);
+		}
+
+		for (j = 0; j < nb_keys; j++) {
+			enc_xor_block((uint64_t *) kdf_out[j], (uint64_t *) result);
+		}
+
+		/* AES CTR */
+		if (cur_salt->version == 1) {
+			enc_xor_block((uint64_t *) ivs[0], (uint64_t *) cur_salt->iv);
+			for (j = 1; j < nb_keys; j++) {
+				enc_xor_block((uint64_t *) ivs[j], (uint64_t *) kdf_out[j]);
+				enc_xor_block((uint64_t *) ivs[j], (uint64_t *) cur_salt->iv);
+			}
+			enc_aes_ctr_iterated(cur_salt->encrypted_data, result, kdf_out[0],
+			                     ivs, AES_BLOCK_SIZE, nb_keys, 1);
+			result_ptr = result + 4;
+			cracked[index] =
+			    (0 == memcmp(result_ptr, "\xd2\xc3\xb4\xa1\x00\x00\x00\x01",
+			                 ENC_SIG_SIZE));
```

Comment:
We'd just write `!memcmp`, and ditto in the other place. More importantly, I suggest you introduce the `any_cracked` logic (grep other formats for it), so that you don't need the loop in `cmp_all`.